In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('diabetes.csv')

In [3]:
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [4]:
df.shape

(768, 9)

In [5]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [6]:
X = df.drop('Outcome',axis=1)
y = df['Outcome']

### Scaling

In [9]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_sc = sc.fit_transform(X)

### Cross Validation 

In [10]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_sc,y,random_state=11,test_size=0.2)

In [12]:
print(X_train.shape)
print(X_test.shape)

(614, 8)
(154, 8)


### KNN  

In [14]:
from sklearn.neighbors import KNeighborsClassifier
clf = KNeighborsClassifier()

In [15]:
clf.fit(X_train,y_train)

KNeighborsClassifier()

In [16]:
clf.score(X_test,y_test)

0.7207792207792207

### K-fold technique 

In [21]:
from sklearn.model_selection import KFold
Kf = KFold(n_splits=5)

In [22]:
# for checking whether i and j are given to training and testing resp.

for i,j in Kf.split(X_sc,y):
    print(i,j)
    break

[154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171
 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189
 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207
 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225
 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243
 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261
 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278 279
 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296 297
 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312 313 314 315
 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332 333
 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348 349 350 351
 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366 367 368 369
 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384 385 386 387
 388 389 390 391 392 393 394 395 396 397 398 399 40

In [20]:
154/614

0.250814332247557

25% of data is used in testing and other 75% of data is used in training

In [23]:
from sklearn.neighbors import KNeighborsClassifier
clf = KNeighborsClassifier()

In [24]:
list1 = []

for i,j in Kf.split(X_sc,y):
    X_train,X_test,y_train,y_test = X_sc[i],X_sc[j],y[i],y[j]
    clf.fit(X_train,y_train)
    score = clf.score(X_test,y_test)
    list1.append(score)

In [25]:
list1

[0.7207792207792207,
 0.6948051948051948,
 0.7532467532467533,
 0.7581699346405228,
 0.7189542483660131]

In [26]:
import numpy as np

In [27]:
accuracy = sum(list1)/5
accuracy

0.7291910703675409

The Accuracy we get is around 72.9% i.e 73% manually

###  Cross Val Score

In [28]:
from sklearn.model_selection import cross_val_score

In [30]:
acc = cross_val_score(clf,X_sc,y,cv=5)
acc

array([0.72077922, 0.73376623, 0.71428571, 0.77124183, 0.7254902 ])

In [32]:
np.mean(acc)

0.733112638994992

The Accuracy we get is 73% 

Instead typing things manually we can use Automatic hyperparameter algorithm

### 1) Gridsearch CV 

In [34]:
from sklearn.svm import SVC
clf = SVC()

In [35]:
d = {'C':[1,2,3,4,5,6,7],'kernel':['rbf','ploy','linear'],'gamma':['auto','scale']}

In [36]:
from sklearn.model_selection import GridSearchCV

In [37]:
gd = GridSearchCV(clf,d,cv=5)

In [38]:
gd.fit(X_sc,y)

c:\users\hp\appdata\local\programs\python\python39\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
70 fits failed out of a total of 210.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
70 fits failed with the following error:
Traceback (most recent call last):
  File "c:\users\hp\appdata\local\programs\python\python39\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\users\hp\appdata\local\programs\python\python39\lib\site-packages\sklearn\svm\_base.py", line 251, in fit
    fit(X, y, sample_weight, solver_type, kernel, random_seed=seed)
  File "c:\users\hp\appdata\local\programs\python\python39\lib\site-pack

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [1, 2, 3, 4, 5, 6, 7], 'gamma': ['auto', 'scale'],
                         'kernel': ['rbf', 'ploy', 'linear']})

In [39]:
gd.cv_results_

{'mean_fit_time': array([0.02168608, 0.00269899, 0.02034292, 0.02102952, 0.        ,
        0.02595797, 0.0257637 , 0.00027037, 0.03437872, 0.02763972,
        0.00173454, 0.02802386, 0.02631178, 0.00312476, 0.04062963,
        0.02235765, 0.        , 0.03935308, 0.02968783, 0.        ,
        0.0414721 , 0.02963824, 0.        , 0.04499431, 0.02593722,
        0.        , 0.05164342, 0.02548881, 0.00335746, 0.05828772,
        0.0259738 , 0.00313759, 0.06031804, 0.02229037, 0.        ,
        0.06013188, 0.02574902, 0.00312505, 0.05512018, 0.02234168,
        0.00312767, 0.06438723]),
 'std_fit_time': array([0.00261566, 0.00539799, 0.00632536, 0.0053309 , 0.        ,
        0.00802615, 0.00723452, 0.00047984, 0.00624412, 0.01231595,
        0.00201285, 0.00557146, 0.00758878, 0.00624952, 0.00640471,
        0.00729198, 0.        , 0.00805744, 0.00863688, 0.        ,
        0.00716392, 0.00737144, 0.        , 0.00658718, 0.00846793,
        0.        , 0.00559846, 0.00704604, 0.006

In [41]:
# convert dictionary into DataFrame
result = pd.DataFrame(gd.cv_results_)
result

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0        0.021686      0.002616         0.010870        0.006377       1   
1        0.002699      0.005398         0.000000        0.000000       1   
2        0.020343      0.006325         0.003543        0.006120       1   
3        0.021030      0.005331         0.016055        0.008843       1   
4        0.000000      0.000000         0.000000        0.000000       1   
5        0.025958      0.008026         0.003124        0.006248       1   
6        0.025764      0.007235         0.010765        0.009141       2   
7        0.000270      0.000480         0.000000        0.000000       2   
8        0.034379      0.006244         0.003124        0.006248       2   
9        0.027640      0.012316         0.016527        0.011740       2   
10       0.001735      0.002013         0.000000        0.000000       2   
11       0.028024      0.005571         0.005386        0.006174       2   
12       0.026312      0.007589         0.012734        0.006383       3   
13       0.003125      0.006250         0.000000        0.000000       3   
14       0.040630      0.006405         0.003526        0.006125       3   
15       0.022358      0.007292         0.013220        0.005530       3   
16       0.000000      0.000000         0.000000        0.000000       3   
17       0.039353      0.008057         0.000000        0.000000       3   
18       0.029688      0.008637         0.009958        0.007740       4   
19       0.000000      0.000000         0.000000        0.000000       4   
20       0.041472      0.007164         0.006335        0.007760       4   
21       0.029638      0.007371         0.009880        0.008080       4   
22       0.000000      0.000000         0.000000        0.000000       4   
23       0.044994      0.006587         0.003324        0.006159       4   
24       0.025937      0.008468         0.008181        0.010472       5   
25       0.000000      0.000000         0.000000        0.000000       5   
26       0.051643      0.005598         0.009596        0.007843       5   
27       0.025489      0.007046         0.012521        0.006260       5   
28       0.003357      0.006715         0.000000        0.000000       5   
29       0.058288      0.008385         0.000000        0.000000       5   
30       0.025974      0.012985         0.016549        0.000787       6   
31       0.003138      0.006275         0.000000        0.000000       6   
32       0.060318      0.005575         0.003487        0.006974       6   
33       0.022290      0.007368         0.016143        0.000566       6   
34       0.000000      0.000000         0.000000        0.000000       6   
35       0.060132      0.006090         0.003390        0.006780       6   
36       0.025749      0.007259         0.009619        0.007866       7   
37       0.003125      0.006250         0.000000        0.000000       7   
38       0.055120      0.008059         0.009375        0.007655       7   
39       0.022342      0.007783         0.015634        0.000023       7   
40       0.003128      0.006255         0.000000        0.000000       7   
41       0.064387      0.010133         0.003332        0.006159       7   

   param_gamma param_kernel                                          params  \
0         auto          rbf      {'C': 1, 'gamma': 'auto', 'kernel': 'rbf'}   
1         auto         ploy     {'C': 1, 'gamma': 'auto', 'kernel': 'ploy'}   
2         auto       linear   {'C': 1, 'gamma': 'auto', 'kernel': 'linear'}   
3        scale          rbf     {'C': 1, 'gamma': 'scale', 'kernel': 'rbf'}   
4        scale         ploy    {'C': 1, 'gamma': 'scale', 'kernel': 'ploy'}   
5        scale       linear  {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}   
6         auto          rbf      {'C': 2, 'gamma': 'auto', 'kernel': 'rbf'}   
7         auto         ploy     {'C': 2, 'gamma': 'auto', 'kernel': 'ploy'}   
8         auto       lin

In [42]:
result[result['rank_test_score']==1]

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
18       0.029688      0.008637         0.009958         0.00774       4   

   param_gamma param_kernel                                      params  \
18        auto          rbf  {'C': 4, 'gamma': 'auto', 'kernel': 'rbf'}   

    split0_test_score  split1_test_score  split2_test_score  \
18            0.75974            0.75974           0.753247   

    split3_test_score  split4_test_score  mean_test_score  std_test_score  \
18           0.816993           0.810458         0.780036        0.027687   

    rank_test_score  
18                1

# The Accuracy is 78% 